crates/rustc_codegen_spirv/src/abi.rs

```diff
@@ -718,21 +718,27 @@ fn trans_aggregate<'tcx>(cx: &CodegenCx<'tcx>, span: Span, ty: TyAndLayout<'tcx>
         FieldsShape::Union(_) => {
             assert!(!ty.is_unsized(), "{ty:#?}");
 
-            // Represent the `union` with its largest case, which should work
-            // for at least `MaybeUninit<T>` (which is between `T` and `()`),
-            // but also potentially some other ones as well.
-            // NOTE(eddyb) even if long-term this may become a byte array, that
-            // only works for "data types" and not "opaque handles" (images etc.).
-            let largest_case = (0..ty.fields.count())
-                .map(|i| ty.field(cx, i))
-                .max_by_key(|case| case.size);
-
-            if let Some(case) = largest_case {
-                assert_eq!(ty.size, case.size);
-                case.spirv_type(span, cx)
+            // If the union has no fields or only one field, represent the `union` with
+            // its largest case, which should work for at least `MaybeUninit<T>` (which
+            // is between `T` and `()`), but also potentially some other ones as well.
+            //
+            // NOTE(eddyb) even if long-term this may become a byte array, that only
+            // works for "data types" and not "opaque handles" (images etc.).
+            if ty.fields.count() <= 1 {
+                let largest_case = (0..ty.fields.count())
+                    .map(|i| ty.field(cx, i))
+                    .max_by_key(|case| case.size);
+
+                if let Some(case) = largest_case {
+                    assert_eq!(ty.size, case.size);
+                    case.spirv_type(span, cx)
+                } else {
+                    assert_eq!(ty.size, Size::ZERO);
+                    create_zst(cx, span, ty)
+                }
```

Comment:
Why not use a match?
```rust
match ty.fields.count() {
    0 => {
        assert_eq!(ty.size, Size::ZERO);
        create_zst(cx, span, ty)
    },
    1 => {
        assert_eq!(ty.size, ty.field(cx, 0).size);
        case.spirv_type(span, cx)
    },
    _ => trans_struct(cx, span, ty),
}
```

Reply:
Because I didn't think of it, good call!